### PR TITLE
85 faultproof withdrawals does not report connection errors metrics correctly

### PR DESCRIPTION
**Pull request by raffaele-oplabs**
**Description**

**Tests**

**Additional context**

**Metadata**

## Summary by CodeRabbit

- **New Features**
	- Introduced a new environment configuration file for the `faultproof_withdrawals` module, enhancing monitoring capabilities.
	- Added detailed documentation on the service's functionalities, including metrics enabling instructions.
	- Enhanced the `OpNodeHelper` with additional client support for L2 operations.
	- Added a new method for retrieving enriched withdrawal events as a map.
	- Introduced a new method to provide string representation for various structs, improving readability.

- **Bug Fixes**
	- Improved event handling and logging mechanisms within the monitoring system.
	- Updated validation logic for withdrawal events to streamline processing.

- **Tests**
	- Added comprehensive test cases for both mainnet and Sepolia environments to validate the `Monitor` functionality.
	- Established a robust testing framework for the `Monitor`, ensuring correct behavior under various conditions.

- **Chores**
	- Simplified the testing framework by removing outdated functions and dependencies.
	- Enhanced the `State` struct for better tracking of withdrawal events and potential issues.
	- Updated dependencies in the `go.mod` file to reflect changes in the library ecosystem.

### CODE REVIEW

File: op-monitorism/faultproof_withdrawals/monitor.go [hunk 1]
```diff
@@ -262,99 +265,76 @@ func (m *Monitor) Run(ctx context.Context) {
...
 			m.log.Error("failed to consume event", "error", err)
-			return nil, err
-		} else if !consumedEvent {
-			newForgeriesGameInProgressEvent = append(newForgeriesGameInProgressEvent, enrichedWithdrawalEvent)
+			return err
 		}
 	}
 
-	return &newForgeriesGameInProgressEvent, nil
+	return nil
```
[the remark this answers is not in the document]
Comment (raffaele-oplabs) [reply op-monitorism/faultproof_withdrawals/monitor.go:301]: This behaviour is on purpose. The goal is to re-process all events if an issue arise.

File: op-monitorism/faultproof_withdrawals/state.go [hunk 2]
```diff
@@ -8,66 +8,153 @@ import (
 	"github.com/ethereum-optimism/optimism/op-service/metrics"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/log"
+	lru "github.com/hashicorp/golang-lru"
 	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+)
+
+const (
+	suspiciousEventsOnChallengerWinsGamesCacheSize = 1000
```
[the remark this answers is not in the document]
Comment (raffaele-oplabs) [reply op-monitorism/faultproof_withdrawals/state.go:17]: For now we keep hardcoded.

File: op-monitorism/faultproof_withdrawals/monitor.go [hunk 3]
```diff
@@ -262,99 +265,74 @@ func (m *Monitor) Run(ctx context.Context) {
...
```
[the remark this answers is not in the document]
Comment (raffaele-oplabs) [reply op-monitorism/faultproof_withdrawals/monitor.go]: order is not important here

File: op-monitorism/faultproof_withdrawals/validator/op_node_helper.go [hunk 4]
```diff
@@ -17,55 +17,80 @@ import (
...
+}
+
+// get latest known L2 block number
+func (on *OpNodeHelper) GetLatestKnownL2BlockNumber() (uint64, error) {
+	LatestKnownL2BlockNumber, err := on.l2OpGethClient.BlockNumber(on.ctx)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get latest known L2 block number: %w", err)
+	}
+	on.LatestKnownL2BlockNumber = LatestKnownL2BlockNumber
+	return LatestKnownL2BlockNumber, nil
```
[the remark this answers is not in the document]
Comment (raffaele-oplabs) [reply op-monitorism/faultproof_withdrawals/validator/op_node_helper.go:66]: The object is not meant to be thread safe.

File: op-monitorism/faultproof_withdrawals/monitor.go [hunk 5]
```diff
@@ -123,14 +124,14 @@ func NewMonitor(ctx context.Context, log log.Logger, m metrics.Factory, cfg CLIC
...
-	state, err := NewState(log, startingL1BlockHeight, latestL1Height)
+	state, err := NewState(log, startingL1BlockHeight, latestL1Height, ret.withdrawalValidator.GetLatestL2Height())
 	if err != nil {
 		return nil, fmt.Errorf("failed to create state: %w", err)
 	}
 	ret.state = *state
 
 	// log state and metrics
-	ret.state.LogState(ret.log)
+	ret.state.LogState()
```
Comment (Ethnical) [inline op-monitorism/faultproof_withdrawals/monitor.go:134]: logState() is not taking a logger  anymore then?

File: op-monitorism/faultproof_withdrawals/state.go [hunk 6]
```diff
@@ -80,23 +170,99 @@ func (s *State) GetPercentages() (uint64, uint64) {
 }
 
 type Metrics struct {
```
Comment (Ethnical) [inline op-monitorism/faultproof_withdrawals/state.go:172]: The whole logic of the Metrics should be moved into the utils.go in the future.

**Review by Ethnical — APPROVED** [review]
Nice new improvements 🚀 
We will need to move more stuff into `utils.go` in the future.

**Pull request merged**